servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DefaultGrpcMetadata.java

```diff
@@ -15,24 +15,76 @@
  */
 package io.servicetalk.grpc.api;
 
+import io.servicetalk.concurrent.internal.DefaultContextMap;
+import io.servicetalk.context.api.ContextMap;
+
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
 import static java.util.Objects.requireNonNull;
 
 class DefaultGrpcMetadata implements GrpcMetadata {
 
     private final String path;
+    private final Supplier<ContextMap> requestContext;
+    private final Supplier<ContextMap> responseContext;
 
     /**
      * Construct a new instance.
      *
      * @param path The path for the associated <a href="https://www.grpc.io">gRPC</a> method.
      */
-    DefaultGrpcMetadata(final String path) {
+    DefaultGrpcMetadata(final String path,
+                        final Supplier<ContextMap> requestContext,
+                        final Supplier<ContextMap> responseContext) {
         this.path = requireNonNull(path, "path");
+        this.requestContext = requireNonNull(requestContext, "requestContext");
+        this.responseContext = requireNonNull(responseContext, "responseContext");
     }
 
     @Deprecated
     @Override
     public String path() {
         return path;
     }
+
+    @Override
+    public ContextMap requestContext() {
+        return requestContext.get();
+    }
+
+    @Override
+    public ContextMap responseContext() {
+        return responseContext.get();
+    }
+
+    boolean responseContext(final ContextMap context) {
+        if (responseContext instanceof LazyContextMapSupplier) {
+            ((LazyContextMapSupplier) responseContext).initialize(context);
+            return true;
+        }
+        return false;
+    }
+
+    static final class LazyContextMapSupplier implements Supplier<ContextMap> {
+
+        @Nullable
+        private ContextMap context;
+
+        @Override
+        public ContextMap get() {
+            if (context == null) {
+                context = new DefaultContextMap();
+            }
+            return context;
+        }
+
+        boolean isInitialized() {
+            return context != null;
+        }
+
+        void initialize(final ContextMap context) {
+            this.context = requireNonNull(context);
```

Comment:
Good questions. Technically, users have access to `grpcMetadata.responseContext()` before they send a request. They can (don't know why) put something there. Overriding it will be surprising. I've added a check to do the swap only if the current `context` is null. Thanks!